protocol/auditlog_test.go

```diff
@@ -0,0 +1,220 @@
+package protocol
+
+import (
+	m "github.com/coniks-sys/coniks-go/merkletree"
+	"testing"
+)
+
+func TestInsertEmptyHistory(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+
+	err := aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history")
+	}
+}
+
+func TestUpdateHistory(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+
+	err := aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history")
+	}
+
+	// update the directory so we can update the audit log
+	d.Update()
+	err = aud.Update("test-server", d.LatestSTR())
+
+	if err != nil {
+		t.Fatal("Error updating the server history")
+	}
+}
+
+func TestInsertPriorHistory(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+
+	// create 10 epochs
+	priorSTRs := make(map[uint64]*m.SignedTreeRoot)
+	for i := 0; i < 10; i++ {
+		priorSTRs[d.LatestSTR().Epoch] = d.LatestSTR()
+		d.Update()
+	}
+
+	// now insert
+	err := aud.Insert("test-server", pk, priorSTRs, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history with prior STRs")
+	}
+}
+
+func TestInsertExistingHistory(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+	err := aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history")
+	}
+
+	// let's make sure that we can't re-insert a new server history into our log
+	err = aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != ErrAuditLog {
+		t.Fatal("Expected an ErrAuditLog when inserting an existing server history")
+	}
+}
+
+func TestUpdateUnknownHistory(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+	err := aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history")
+	}
+
+	// let's make sure that we can't re-insert a new server history into our log
+	err = aud.Update("unknown", d.LatestSTR())
+	if err != ErrAuditLog {
+		t.Fatal("Expected an ErrAuditLog when updating an unknown server history")
+	}
+}
+
+func TestGetObservedSTR(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+	err := aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history")
+	}
+
+	res, err := aud.GetObservedSTR(&AuditingRequest{
+		DirectoryAddr: "test-server"})
+	obs := res.DirectoryResponse.(*ObservedSTR)
+	if err != ReqSuccess {
```

Comment:
Actually, I don't think that would work because I use `res` below for other operations.